src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java

```diff
@@ -48,81 +51,133 @@
+        String localDc = getDatacenterSource.apply(localAddress);
 
+        peers.remove(localAddress);
         if (peers.isEmpty())
             return true;
 
-        logger.info("choosing to block until {}% of the {} known peers are marked alive and connections are established; max time to wait = {} seconds",
-                    targetPercent, peers.size(), TimeUnit.NANOSECONDS.toSeconds(timeoutNanos));
+        // make a copy of the datacenter mapping (in case gossip updates happen during this method or some such)
+        Map<InetAddressAndPort, String> datacenterMap = peers.stream()
+                                                             .collect(Collectors.toMap(k -> k, getDatacenterSource));
+        Function<InetAddressAndPort, String> getDatacenter = datacenterMap::get;
 
-        long startNanos = System.nanoTime();
+        Map<String, Set<InetAddressAndPort>> peersByDc = peers.stream()
+                                                              .collect(Collectors.groupingBy(getDatacenter,
+                                                                                             Collectors.toSet()));
+
+        if (!blockForRemoteDcs)
+        {
+            peersByDc.keySet().retainAll(Collections.singleton(localDc));
+            logger.info("Blocking coordination until only a single peer is DOWN in the local datacenter, timeout={}s",
+                        TimeUnit.NANOSECONDS.toSeconds(timeoutNanos));
+        }
+        else
+        {
+            logger.info("Blocking coordination until only a single peer is DOWN in each datacenter, timeout={}s",
+                        TimeUnit.NANOSECONDS.toSeconds(timeoutNanos));
+        }
 
         AckMap acks = new AckMap(3);
-        int target = (int) ((targetPercent / 100.0) * peers.size());
-        CountDownLatch latch = new CountDownLatch(target);
+        Map<String, CountDownLatch> latchMap = new HashMap<>(peersByDc.size());
+        for (Map.Entry<String, Set<InetAddressAndPort>> entry: peersByDc.entrySet())
+        {
+            latchMap.put(entry.getKey(), new CountDownLatch(Math.max(entry.getValue().size() - 1, 0)));
+        }
+
+        long startNanos = System.nanoTime();
 
         // set up a listener to react to new nodes becoming alive (in gossip), and account for all the nodes that are already alive
-        Set<InetAddressAndPort> alivePeers = Sets.newSetFromMap(new ConcurrentHashMap<>());
-        AliveListener listener = new AliveListener(alivePeers, latch, acks);
+        Set<InetAddressAndPort> alivePeers = Collections.newSetFromMap(new ConcurrentHashMap<>());
+        AliveListener listener = new AliveListener(alivePeers, latchMap, acks, getDatacenter);
         Gossiper.instance.register(listener);
 
         // send out a ping message to open up the non-gossip connections
-        sendPingMessages(peers, latch, acks);
+        sendPingMessages(peers, latchMap, acks, getDatacenter);
```

Comment:
Comment that pings are sent to all peers even if latchMap doesn't contain a latch for them.

Reply:
Done.